File: client-data/src/test/java/com/jraska/github/client/users/UserDetailViewModelTest.java
package com.jraska.github.client.users;

import com.jraska.github.client.FakeConfig;
import com.jraska.github.client.Navigator;
import com.jraska.github.client.analytics.EventAnalytics;
import com.jraska.github.client.rx.AppSchedulers;

import org.junit.Test;
import org.threeten.bp.LocalDateTime;

import java.util.Collections;

import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class UserDetailViewModelTest {
  @Test
  public void whenSameLoginMultipleTimes_thenOnlyOneObservableCreated() {
    UsersRepository usersRepository = mock(UsersRepository.class);


    Observable<UserDetail> detailObservable = Observable.just(testDetail());
    when(usersRepository.getUserDetail(any(), anyInt())).thenReturn(detailObservable);

    FakeConfig config = FakeConfig.create("user_detail_section_size", 3L);

    UserDetailViewModel viewModel = new UserDetailViewModel(usersRepository,
      trampoline(), mock(Navigator.class), EventAnalytics.EMPTY, config);

    String login = "someLogin";

    viewModel.userDetail(login);
    verify(usersRepository).getUserDetail(any(), eq(3));

    viewModel.userDetail(login);
    verify(usersRepository).getUserDetail(any(), anyInt());

    viewModel.userDetail("different" + login);
    verify(usersRepository, times(2)).getUserDetail(any(), eq(3));
  }

  public static AppSchedulers trampoline() {
    return new AppSchedulers(Schedulers.trampoline(), Schedulers.trampoline(),
      Schedulers.trampoline());
  }

  static UserDetail testDetail() {
    User user = new User("login", "url", true, "url");
    UserStats stats = new UserStats(0, 0, 0, LocalDateTime.MIN);
    return new UserDetail(user, stats, Collections.emptyList(), Collections.emptyList());
  }
}

File: client-data/src/test/java/com/jraska/github/client/users/UserDetailViewModelTest.kt
package com.jraska.github.client.users

import android.arch.core.executor.testing.InstantTaskExecutorRule
import com.jraska.github.client.FakeConfig
import com.jraska.github.client.Navigator
import com.jraska.github.client.analytics.EventAnalytics
import com.jraska.github.client.rx.AppSchedulers
import com.jraska.livedata.test
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.Mockito.*
import org.threeten.bp.Instant

class UserDetailViewModelTest {
  @get:Rule val testRule = InstantTaskExecutorRule()

  lateinit var usersRepository: UsersRepository

  private lateinit var viewModel: UserDetailViewModel

  @Before
  fun before() {
    usersRepository = mock(UsersRepository::class.java)

    val detailObservable = Observable.just(testDetail())
    `when`(usersRepository.getUserDetail("someLogin", 3)).thenReturn(detailObservable)
    `when`(usersRepository.getUserDetail("different", 3)).thenReturn(detailObservable)
    val config = FakeConfig.create(mapOf("user_detail_section_size" to 3L))

    viewModel = UserDetailViewModel(usersRepository,
      trampoline(), mock(Navigator::class.java), EventAnalytics.EMPTY, config)
  }

  @Test
  fun whenLiveData_thenCorrectUser() {
    viewModel.userDetail("someLogin")
      .test()
      .assertHasValue()
      .assertValue { it.isLoading.not() }
      .assertValue { it.error() == null }
  }

  @Test
  fun whenSameLoginMultipleTimes_thenOnlyOneObservableCreated() {
    val login = "someLogin"

    viewModel.userDetail(login).test().dispose()
    verify(usersRepository).getUserDetail("someLogin", 3)

    viewModel.userDetail(login).test().dispose()
    verify(usersRepository).getUserDetail("someLogin", 3)

    viewModel.userDetail("different").test().dispose()
    verify(usersRepository).getUserDetail("different", 3)
  }

  companion object {
    internal fun testDetail(): UserDetail {
      val user = User("someLogin", "url", true)
      val stats = UserStats(0, 0, 0, Instant.MIN)
      return UserDetail(user, stats, emptyList(), emptyList())
    }
  }
}

fun trampoline(): AppSchedulers {
  return AppSchedulers(Schedulers.trampoline(), Schedulers.trampoline(),
    Schedulers.trampoline())
}

